jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,702 @@
+{
+  Set_ = std::make_unique<PointerObjectSet>();
+  Constraints_ = std::make_unique<PointerObjectConstraintSet>(*Set_);
+
+  AnalyzeRvsdg(module.Rvsdg());
+
+  Constraints_->Solve();
+
+  auto result = ConstructPointsToGraphFromPointerObjectSet(*Set_);
+
+  Constraints_.reset();
+  Set_.reset();
+
+  return result;
+}
+
+std::unique_ptr<PointsToGraph>
+Andersen::Analyze(const RvsdgModule & module)
+{
+  util::StatisticsCollector statisticsCollector;
+  return Analyze(module, statisticsCollector);
+}
+
+std::unique_ptr<PointsToGraph>
+Andersen::ConstructPointsToGraphFromPointerObjectSet(const PointerObjectSet & set)
+{
+  auto pointsToGraph = PointsToGraph::Create();
+
+  // memory nodes are the nodes that can be pointed to in the points-to graph.
+  // This vector has the same indexing as the nodes themselves, register nodes become nullptr.
+  std::vector<PointsToGraph::MemoryNode *> memoryNodes(set.NumPointerObjects());
+
+  // Nodes that should point to external in the final graph.
+  // They also get explicit edges connecting them to all escaped memory nodes.
+  std::vector<PointsToGraph::Node *> pointsToExternal;
+
+  // A list of all memory nodes that have been marked as escaped
+  std::vector<PointsToGraph::MemoryNode *> escapedMemoryNodes;
+
+  // First all memory nodes are created
+  for (auto [allocaNode, pointerObjectIndex] : set.GetAllocaMap())
+  {
+    auto & node = PointsToGraph::AllocaNode::Create(*pointsToGraph, *allocaNode);
+    memoryNodes[pointerObjectIndex] = &node;
+  }
+  for (auto [mallocNode, pointerObjectIndex] : set.GetMallocMap())
+  {
+    auto & node = PointsToGraph::MallocNode::Create(*pointsToGraph, *mallocNode);
+    memoryNodes[pointerObjectIndex] = &node;
+  }
+  for (auto [deltaNode, pointerObjectIndex] : set.GetGlobalMap())
+  {
+    auto & node = PointsToGraph::DeltaNode::Create(*pointsToGraph, *deltaNode);
+    memoryNodes[pointerObjectIndex] = &node;
+  }
+  for (auto [lambdaNode, pointerObjectIndex] : set.GetFunctionMap())
+  {
+    auto & node = PointsToGraph::LambdaNode::Create(*pointsToGraph, *lambdaNode);
+    memoryNodes[pointerObjectIndex] = &node;
+  }
+  for (auto [argument, pointerObjectIndex] : set.GetImportMap())
+  {
+    auto & node = PointsToGraph::ImportNode::Create(*pointsToGraph, *argument);
+    memoryNodes[pointerObjectIndex] = &node;
+  }
+
+  // Helper function for attaching PointsToGraph nodes to their pointees, based on the
+  // PointerObject's points-to set.
+  auto applyPointsToSet = [&](PointsToGraph::Node & node, PointerObject::Index index)
+  {
+    // Add all PointsToGraph nodes who should point to external to the list
+    if (set.GetPointerObject(index).PointsToExternal())
+      pointsToExternal.push_back(&node);
+
+    for (const auto targetIdx : set.GetPointsToSet(index).Items())
+    {
+      // Only PointerObjects corresponding to memory nodes can be members of points-to sets
+      JLM_ASSERT(memoryNodes[targetIdx]);
```

Comment:
This check does not do anything, no? Should it have been `JLM_ASSERT(targetIdx < memoryNodes.size())`?

Reply:
It does, memoryNodes is nullptr for all indexes that correspond to PointerObjects of Register kind